pkg/kv/kvserver/store.go

```diff
@@ -2870,15 +2871,44 @@ func (s *Store) checkpointsDir() string {
 	return filepath.Join(s.engine.GetAuxiliaryDir(), "checkpoints")
 }
 
-// checkpoint creates a RocksDB checkpoint in the auxiliary directory with the
-// provided tag used in the filepath. The filepath for the checkpoint directory
-// is returned.
-func (s *Store) checkpoint(ctx context.Context, tag string) (string, error) {
+// checkpoint creates a Pebble checkpoint in the auxiliary directory with the
+// provided tag used in the filepath. Returns the path to the created checkpoint
+// directory.
+//
+// Creates the checkpoint of the entire storage if the provided range descriptor
+// is nil. Otherwise, creates a "narrow" checkpoint that includes the provided
+// range, and possibly a few neighbouring ranges.
+func (s *Store) checkpoint(tag string, desc *roachpb.RangeDescriptor) (string, error) {
 	checkpointBase := s.checkpointsDir()
 	_ = s.engine.MkdirAll(checkpointBase)
-
 	checkpointDir := filepath.Join(checkpointBase, tag)
-	if err := s.engine.CreateCheckpoint(checkpointDir); err != nil {
+
+	var spans []pebble.CheckpointSpan
+	if desc != nil {
+		// Find left and right neighbour of the range.
+		s.mu.RLock()
+		left := s.mu.replicasByKey.LookupPrecedingReplica(context.Background(), desc.StartKey).Desc()
+		var right *roachpb.RangeDescriptor
+		err := s.mu.replicasByKey.VisitKeyRange(context.Background(), desc.EndKey, roachpb.RKeyMax,
```

Comment:
`LookupPrecedingReplica` finds a `Replica`, whereas the `VisitKeyRange` with the visitor written here find a `Replica` or a placeholder. We could unify this so that both `left` and `right` are `Replica`, or both are `Replica`/placeholder. Which way makes more sense?

Placeholders don't have much data, so could be skipped until we find a regular `Replica`. OTOH, if there are too many placeholders (or all of them), we would still checkpoint a large space.

Reply:
To guard against checkpointing too much, we can provide 3 `CheckpointSpan`s below, instead of a single wider one. The risk with one wider span is that the neighbouring ranges are not locked, so we may end up touching more ranges than intended.

Reply:
>LookupPrecedingReplica finds a Replica, whereas the VisitKeyRange with the visitor written here finds a Replica or a placeholder. We could unify this so that both left and right are Replica, or both are Replica/placeholder. Which way makes more sense?

We don't care about placeholders, I think. We can e.g. add a corresponding `LookupNextReplica` with the same behavior.

>To guard against checkpointing too much, we can provide 3 CheckpointSpans below, instead of a single wider one. The risk with one wider span is that the neighbouring ranges are not locked, so we may end up touching more ranges than intended.

Yeah, this can be pathological in certain extreme cases, e.g. where we recently removed a bunch of replicas but they haven't been compacted all the way down yet. But I suppose there's the same risk if we just merged a bunch of ranges into the inconsistent range.

I think we probably want to checkpoint the entire span between the neighbouring replicas, because any recently-removed ranges could be relevant to the inconsistency, and we'll just have to take the risk that they may contain more than 3 ranges worth of data. In most cases they wouldn't, and in any case it's a strict improvement over the current state -- we can reconsider if we do find any issues with it later.

Reply:
Done.